src/lib/market-service/portals/portals.ts

```diff
@@ -0,0 +1,380 @@
+  decimals: number
+  symbol: string
+  address: string
+  images: string[] | undefined
+  pricePerShare: string | undefined
+}
+
+type GetTokensResponse = {
+  totalItems: number
+  pageItems: number
+  more: boolean
+  page: number
+  tokens: TokenInfo[]
+}
+
+type HistoryResponse = {
+  totalItems: number
+  pageItems: number
+  more: boolean
+  page: number
+  history: {
+    time: string
+    highPrice: string
+    lowPrice: string
+    openPrice: string
+    closePrice: string
+    liquidity: string
+    reserves: string[]
+    totalSupply: string
+    pricePerShare: string
+    volume1dUsd: string
+    apy: string
+  }[]
+}
+
+const CHAIN_ID_TO_PORTALS_NETWORK: Partial<Record<ChainId, string>> = {
+  [avalancheChainId]: 'avalanche',
+  [ethChainId]: 'ethereum',
+  [polygonChainId]: 'polygon',
+  [bscChainId]: 'bsc',
+  [optimismChainId]: 'optimism',
+  [arbitrumChainId]: 'arbitrum',
+  [gnosisChainId]: 'gnosis',
+  [baseChainId]: 'base',
+}
+
+// TODO(gomes): can we have root-level utils so we don't redeclare this from the scripts folder?
+const { throttle, clear } = (({
+  capacity,
+  costPerReq,
+  drainPerInterval,
+  intervalMs,
+}: {
+  capacity: number
+  costPerReq: number
+  drainPerInterval: number
+  intervalMs: number
+}) => {
+  let currentLevel = 0
+  let pendingResolves: ((value?: unknown) => void)[] = []
+
+  const drain = () => {
+    const drainAmount = Math.min(currentLevel, drainPerInterval)
+    currentLevel -= drainAmount
+
+    // Resolve pending promises if there's enough capacity
+    while (pendingResolves.length > 0 && currentLevel + costPerReq <= capacity) {
+      const resolve = pendingResolves.shift()
+      if (resolve) {
+        currentLevel += costPerReq
+        resolve()
+      }
+    }
+  }
+
+  // Start the interval to drain the capacity
+  const intervalId = setInterval(drain, intervalMs)
+
+  const throttle = async () => {
+    if (currentLevel + costPerReq <= capacity) {
+      // If adding another request doesn't exceed capacity, proceed immediately
+      currentLevel += costPerReq
+    } else {
+      // Otherwise, wait until there's enough capacity
+      await new Promise(resolve => {
+        pendingResolves.push(resolve)
+      })
+    }
+  }
+
+  const clear = () => clearInterval(intervalId)
+
+  return { throttle, clear }
+})({
+  capacity: 500, // 500 rpm as per https://github.com/shapeshift/web/pull/7401#discussion_r1687499650
+  costPerReq: 1,
+  drainPerInterval: 125, // Replenish 125 requests every 15 seconds
+  intervalMs: 15000, // 15 seconds
+})
+
+const axios = setupCache(Axios.create(), { ttl: DEFAULT_CACHE_TTL_MS, cacheTakeover: false })
+
+const PORTALS_API_KEY = process.env.REACT_APP_PORTALS_API_KEY
+
+export class PortalsMarketService implements MarketService {
+  baseUrl = 'https://api.portals.fi'
+
+  private readonly defaultGetByMarketCapArgs: FindAllMarketArgs = {
+    count: 250,
+  }
+
+  async findAll(args?: FindAllMarketArgs): Promise<MarketCapResult> {
```

Comment:
Added for the sake of completeness and satisfying `MarketService`, though we could just as well return an empty object and call it a day. 

Given the fact that Portals is *after* coingecko and Coincap in the market-service providers, this is realistically never going to be hit.

However, in the unlikely scenario where both Coingecko and Coincap were to be down, we'd currently be left without any market data initial fetch (i.e initial batches) and no subsequent market-data when navigating to asset page either.

This brings a free redundancy path while being a complete implementation of a provider 🎉 

Tested locally after a full cache clear and Portals prioritized over all other providers, and confirmed market-data still looks sane for initial batches fetch.

Reply:
That's cool, moar redundancy!